File: android/app/src/main/java/com/izzatismail/midtrim/presentation/viewmodel/VideoSelectionViewModel.kt
package com.izzatismail.midtrim.presentation.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.izzatismail.midtrim.domain.entity.VideoMetadata
import com.izzatismail.midtrim.domain.usecase.CalculateMergedDurationUseCase
import com.izzatismail.midtrim.domain.usecase.FetchEntitlementStatusUseCase
import com.izzatismail.midtrim.domain.usecase.ImportVideosUseCase
import com.izzatismail.midtrim.domain.usecase.ReorderVideosUseCase
import com.izzatismail.midtrim.domain.usecase.ValidateTrimDurationUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class VideoSelectionUiState(
    val selectedVideos: List<VideoMetadata> = emptyList(),
    val trimDuration: Double = 1.0,
    val mergedDuration: Double = 0.0,
    val isPaidUser: Boolean = false,
    val isLoading: Boolean = false,
    val importError: String? = null
)

class VideoSelectionViewModel(
    private val importVideosUseCase: ImportVideosUseCase,
    private val reorderVideosUseCase: ReorderVideosUseCase,
    private val calculateMergedDurationUseCase: CalculateMergedDurationUseCase,
    private val validateTrimDurationUseCase: ValidateTrimDurationUseCase,
    private val fetchEntitlementStatusUseCase: FetchEntitlementStatusUseCase
) : ViewModel() {

    private val _uiState = MutableStateFlow(VideoSelectionUiState())
    val uiState: StateFlow<VideoSelectionUiState> = _uiState.asStateFlow()

    private val existingUris: Set<String>
        get() = _uiState.value.selectedVideos.map { it.uri }.toSet()

    fun initialize() {
        viewModelScope.launch {
            val isPaid = fetchEntitlementStatusUseCase.isPaidUser
            _uiState.value = _uiState.value.copy(isPaidUser = isPaid)
        }
    }

    fun importVideos(videoUris: List<String>) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, importError = null)
            try {
                val metadata = importVideosUseCase.execute(videoUris)
                val existing = existingUris
                val newMetadata = metadata.filter { it.uri !in existing }
                val all = _uiState.value.selectedVideos + newMetadata
                val merged = calculateMergedDurationUseCase.execute(
                    _uiState.value.trimDuration, all.size
                )
                _uiState.value = _uiState.value.copy(
                    selectedVideos = all,
                    mergedDuration = merged,
                    isLoading = false
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    importError = e.message
                )
            }
        }
    }

    fun removeVideo(index: Int) {
        val updated = _uiState.value.selectedVideos.toMutableList()
        if (index in updated.indices) {
            updated.removeAt(index)
            val merged = calculateMergedDurationUseCase.execute(
                _uiState.value.trimDuration, updated.size
            )
            _uiState.value = _uiState.value.copy(
                selectedVideos = updated,
                mergedDuration = merged
            )
        }
    }

    fun reorder(fromIndex: Int, toIndex: Int) {
        val items = _uiState.value.selectedVideos.toMutableList()
        if (fromIndex in items.indices && toIndex in items.indices) {
            val item = items.removeAt(fromIndex)
            items.add(toIndex, item)
            _uiState.value = _uiState.value.copy(selectedVideos = items)
        }
    }

    fun setTrimDuration(duration: Double) {
        if (validateTrimDurationUseCase.isAllowed(duration, _uiState.value.isPaidUser)) {
            val merged = calculateMergedDurationUseCase.execute(
                duration, _uiState.value.selectedVideos.size
            )
            _uiState.value = _uiState.value.copy(
                trimDuration = duration,
                mergedDuration = merged
            )
        }
    }

    fun clearImportError() {
        _uiState.value = _uiState.value.copy(importError = null)
    }
}

File: ios/MidTrim/MidTrim/Presentation/ViewModels/VideoSelectionViewModel.swift
import SwiftUI
import Combine

@MainActor
class VideoSelectionViewModel: ObservableObject {
    @Published var uiState = VideoSelectionUiState()

    private let importVideosUseCase: ImportVideosUseCase
    private let reorderVideosUseCase: ReorderVideosUseCase
    private let calculateMergedDurationUseCase: CalculateMergedDurationUseCase
    private let validateTrimDurationUseCase: ValidateTrimDurationUseCase
    private let fetchEntitlementStatusUseCase: FetchEntitlementStatusUseCase

    init(
        importVideosUseCase: ImportVideosUseCase,
        reorderVideosUseCase: ReorderVideosUseCase,
        calculateMergedDurationUseCase: CalculateMergedDurationUseCase,
        validateTrimDurationUseCase: ValidateTrimDurationUseCase,
        fetchEntitlementStatusUseCase: FetchEntitlementStatusUseCase
    ) {
        self.importVideosUseCase = importVideosUseCase
        self.reorderVideosUseCase = reorderVideosUseCase
        self.calculateMergedDurationUseCase = calculateMergedDurationUseCase
        self.validateTrimDurationUseCase = validateTrimDurationUseCase
        self.fetchEntitlementStatusUseCase = fetchEntitlementStatusUseCase
    }

    func initialize() {
        Task { @MainActor in
            uiState.isPaidUser = await fetchEntitlementStatusUseCase.isPaidUser
        }
    }

    func importVideos(_ uris: [String]) {
        Task { @MainActor in
            uiState.isLoading = true
            do {
                let metadata = try await importVideosUseCase.execute(videoURIs: uris)
                uiState.selectedVideos.append(contentsOf: metadata)
                uiState.mergedDuration = calculateMergedDurationUseCase.execute(
                    trimDuration: uiState.trimDuration,
                    videoCount: uiState.selectedVideos.count
                )
                uiState.isLoading = false
            } catch {
                uiState.isLoading = false
                uiState.importError = error.localizedDescription
            }
        }
    }

    func removeVideo(at index: Int) {
        guard uiState.selectedVideos.indices.contains(index) else { return }
        uiState.selectedVideos.remove(at: index)
        uiState.mergedDuration = calculateMergedDurationUseCase.execute(
            trimDuration: uiState.trimDuration,
            videoCount: uiState.selectedVideos.count
        )
    }

    func reorder(from source: IndexSet, to destination: Int) {
        uiState.selectedVideos.move(fromOffsets: source, toOffset: destination)
    }

    func setTrimDuration(_ duration: Double) {
        guard validateTrimDurationUseCase.isAllowed(
            trimDuration: duration,
            isPaidUser: uiState.isPaidUser
        ) else { return }
        uiState.trimDuration = duration
        uiState.mergedDuration = calculateMergedDurationUseCase.execute(
            trimDuration: duration,
            videoCount: uiState.selectedVideos.count
        )
    }
}

struct VideoSelectionUiState {
    var selectedVideos: [VideoMetadata] = []
    var trimDuration: Double = 1.0
    var mergedDuration: Double = 0.0
    var isPaidUser = false
    var isLoading = false
    var importError: String?
}
